src/Draco.Compiler.Internal/Utilities/Result.cs

```diff
@@ -0,0 +1,175 @@
+        : this.error;
+
+    /// <summary>
+    /// Whether the result is an ok value or not.
+    /// </summary>
+    [MemberNotNullWhen(true, nameof(Ok))]
+    [MemberNotNullWhen(false, nameof(Error))]
+    public bool IsOk =>
+        this.isOk;
+
+    /// <summary>
+    /// Initializes a new <see cref="Result{T, TError}"/> with an ok value.
+    /// </summary>
+    /// <param name="ok">The ok value.</param>
+    public Result(TOk ok)
+    {
+        this.ok = ok;
+        this.error = default!;
+        this.isOk = true;
+    }
+
+    /// <summary>
+    /// Initializes a new <see cref="Result{T, TError}"/> with an error value.
+    /// </summary>
+    /// <param name="error">The error value.</param>
+    public Result(TError error)
+    {
+        this.ok = default!;
+        this.error = error;
+        this.isOk = false;
+    }
+
+    /// <summary>
+    /// Maps the ok value of the result to another type.
+    /// </summary>
+    /// <typeparam name="UOk">The type of the new ok value.</typeparam>
+    /// <param name="transform">A transformation function mapping the ok value.</param>
+    /// <returns>A new result of either the new transformed ok value or the previous error.</returns>
+    public Result<UOk, TError> MapOk<UOk>(Func<TOk, UOk> transform) => this.isOk
+        ? new(transform(this.ok))
+        : new(this.error);
+
+    /// <summary>
+    /// Maps the inner error value of the result to another error type.
+    /// </summary>
+    /// <typeparam name="UError">The type of the new error value.</typeparam>
+    /// <param name="transform">A transformation function mapping the error value.</param>
+    /// <returns>A new result of either the new transformed error value or the previous ok value.</returns>
+    public Result<TOk, UError> MapError<UError>(Func<TError, UError> transform) => this.isOk
+        ? new(this.ok)
+        : new(transform(this.error));
+
+    /// <summary>
+    /// Binds the ok value to a new result with another ok value type.
+    /// </summary>
+    /// <typeparam name="UOk">The type of the new ok value.</typeparam>
+    /// <param name="transform">A transformation function mapping the ok value to a new result.</param>
+    /// <returns>A new result of either the returned result or the previous error.</returns>
+    public Result<UOk, TError> BindOk<UOk>(Func<TOk, Result<UOk, TError>> transform) => this.isOk
+        ? transform(this.ok)
+        : new(this.error);
+
+    /// <summary>
+    /// Binds the error value to a new result with another error type.
+    /// </summary>
+    /// <typeparam name="UError">The type of the new error value.</typeparam>
+    /// <param name="transform">A transformation function mapping the error value to a new result.</param>
+    /// <returns>A new result of either the returned result or the previous ok value.</returns>
+    public Result<TOk, UError> BindError<UError>(Func<TError, Result<TOk, UError>> transform) =>
+        this.isOk
```

Comment:
Missed ternary, or is the line too long?

Reply:
missed